pkg/controller/elasticsearch/nodespec/statefulset.go

```diff
@@ -116,21 +123,39 @@ func BuildStatefulSet(
 
 func setVolumeClaimsControllerReference(
 	persistentVolumeClaims []corev1.PersistentVolumeClaim,
+	existingClaims []corev1.PersistentVolumeClaim,
 	es v1beta1.Elasticsearch,
 	scheme *runtime.Scheme,
 ) ([]corev1.PersistentVolumeClaim, error) {
 	// set the owner reference of all volume claims to the ES resource,
 	// so PVC get deleted automatically upon Elasticsearch resource deletion
 	claims := make([]corev1.PersistentVolumeClaim, 0, len(persistentVolumeClaims))
 	for _, claim := range persistentVolumeClaims {
-		// Set the claim namespace to match the ES namespace.
-		// This is technically not required, but `SetControllerReference` does a safety check on
-		// object vs. owner namespace mismatch. We know the PVC will end up in ES namespace anyway,
-		// so it's safe to include it.
+		if existingClaim := getClaimMatchingName(existingClaims, claim.Name); existingClaim != nil {
+			// This claim already exists in the actual resource. Since the volumeClaimTemplates section of
+			// a StatefulSet is immutable, any modification to it will be rejected in the StatefulSet update.
+			// This is fine and we let it error-out. It is caught in a user-friendly way by the validating webhook.
+			//
+			// However, there is one case where the claim we build may differ from the existing one, that was
+			// built with a prior version of the operator. If the Elasticsearch apiVersion has changed,
+			// from eg. `v1beta1` to `v1`, we want to keep the existing ownerRef (pointing to eg. a `v1beta1` owner).
+			// Having ownerReferences with a "deprecated" apiVersion is fine, and does not prevent resources
+			// to be garbage collected as expected.
+			claim.OwnerReferences = existingClaim.OwnerReferences
+
+			claims = append(claims, claim)
+			continue
+		}
+
+		// Temporarily set the claim namespace to match the ES namespace, then set it back to empty.
+		// `SetControllerReference` does a safety check on object vs. owner namespace mismatch to cover common errors,
+		// but in this particular case we don't need to set a namespace in the claim template.
 		claim.Namespace = es.Namespace
 		if err := reconciler.SetControllerReference(&es, &claim, scheme); err != nil {
 			return nil, err
 		}
+		claim.Namespace = ""
```

Comment:
This deals with https://github.com/elastic/cloud-on-k8s/pull/2211/commits/bdf1e1f3da641224e97eff6cf6030526f9707603 in a backward-compatible way.

Reply:
Thanks for including the explanation. Since we're using our own version for now it would have taken me a bit to understand